From 9f283471c86a689ca8f7a518556045d47d2637f3 Mon Sep 17 00:00:00 2001
From: Thirusenthil Kumaran J <thirusen@qti.qualcomm.com>
Date: Thu, 26 Jun 2025 20:08:28 +0530
Subject: [PATCH 5/5] UPSTREAM:PROTOCOL:wifi:ath12k: Intersect AFC payload for
 dual 6 GHz radio

Currently, if two 6 GHz radios are present in the same HW group,
the AFC payload received for one radio overwrites the AFC payload
previously received for the other radio. This can lead to incorrect
SP rules being stored in the wiphy regd.

To fix this, intersect the AFC payloads for both radios and combine
the SP rules from both radios into a single set of rules.
This ensures that the final set of SP rules reflects the combined
payloads from both radios, preventing any overwriting issues.

Currently, when a new AFC payload is received in host, the memory for
old AFC payload stored in struct ath12k is cleared and a new memory
is allocated and assigned directly to the ath12k data structure.

This causes concurrent access issues where there is a possibility of
the AFC payload being cleared for a radio, while the same is being
accessed by a partner 6 GHz radio as part of SP rules intersection.

To prevent the concurrent access issues, do not free the AFC payload
stored in struct ath12k when a new AFC payload is received. Also,
allocate a new AFC payload and pass the payload to regulatory
API for validation, upon which the AFC payload is assigned to
the ath12k data structure.

To ensure data consistency and to avoid concurrent access to AFC
payloads of both the radios, take an 'ah' afc lock to serialize the
AFC payload processing in the regulatory handler.
Free the AFC payload for a radio under a radio level lock to
avoid concurrent access to the AFC payload when it is being freed.

Make changes to the prototype of the regulatory handler to
accept the AFC payload as an argument.

Patch-work: None
Tag: 6G_REGULATORY
Feature: AFC
Patch-dependency:
subsys/0048-cfg80211-mac80211-regulatory-awgn-and-6g-channel-cha.patch
ath12k/0346-02-wifi-ath12k-Add-support-for-processing-AFC-power-eve.patch
ath12k/0346-06-wifi-ath12k-Extract-AFC-response-and-send-AFC-cmd-to.patch
ath12k/0380-a-UPSTREAM-PROTOCOL-wifi-ath12k-AFC-payload-cleanup-an.patch
ath12k/0380-b-UPSTREAM-PROTOCOL-wifi-ath12k-Generate-SP-rules-base.patch
ath12k/0380-c-UPSTREAM-PROTOCOL-wifi-ath12k-Generate-NO_IR-rules-b.patch

Signed-off-by: Thirusenthil Kumaran J <thirusen@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/core.h   |   4 +
 drivers/net/wireless/ath/ath12k/mac.c    |   1 +
 drivers/net/wireless/ath/ath12k/reg.c    | 220 +++++++++++++++++++++--
 drivers/net/wireless/ath/ath12k/reg.h    |  14 +-
 drivers/net/wireless/ath/ath12k/vendor.c |  24 +--
 drivers/net/wireless/ath/ath12k/vendor.h |   7 +-
 drivers/net/wireless/ath/ath12k/wmi.c    |  29 +--
 7 files changed, 260 insertions(+), 39 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index e653a794e94d..b36ce94eecaf 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -1133,6 +1133,10 @@ struct ath12k_hw {
 	enum ath12k_hw_state state;
 	bool regd_updated;
 	bool use_6ghz_regd;
+	/* Protect concurrent access to the wiphy regd when processing AFC
+	 * power event from multple radios.
+	 */
+	spinlock_t afc_lock;
 
 	u8 num_radio;
 
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 3bd7ea421fce..168b6a3e75d4 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -19393,6 +19393,7 @@ static struct ath12k_hw *ath12k_mac_hw_allocate(struct ath12k_hw_group *ag,
 
 	mutex_init(&ah->hw_mutex);
 
+	spin_lock_init(&ah->afc_lock);
 	spin_lock_init(&ah->dp_hw.peer_lock);
 	INIT_LIST_HEAD(&ah->dp_hw.peers);
 
diff --git a/drivers/net/wireless/ath/ath12k/reg.c b/drivers/net/wireless/ath/ath12k/reg.c
index e157c33dd561..ec04031b77a5 100644
--- a/drivers/net/wireless/ath/ath12k/reg.c
+++ b/drivers/net/wireless/ath/ath12k/reg.c
@@ -1763,6 +1763,181 @@ ath12k_reg_generate_valid_afc_ranges(struct ath12k *ar,
 	return 0;
 }
 
+/**
+ * ath12k_concat_sorted_ranges() - Concatenate two sorted AFC ranges
+ * @ab: pointer to ath12k_base
+ * @this_radio_6ghz_ranges: Input / Output pointer to the AFC frequency ranges
+ * @n_this_6ghz_ranges: Input / Output pointer to the number of AFC ranges
+ * @partner_radio_6ghz_ranges: Pointer to the partner radio AFC frequency ranges
+ * @n_partner_6ghz_ranges: Number of partner radio AFC frequency ranges
+ *
+ * This API concatenates the sorted AFC frequency ranges from both radios.
+ *
+ * Return: 0 on success, negative error code on failure
+ */
+static int
+ath12k_concat_sorted_ranges(struct ath12k_base *ab,
+			    struct ath12k_afc_freq_obj **this_radio_6ghz_ranges,
+			    u16 *n_this_6ghz_ranges,
+			    const struct ath12k_afc_freq_obj *partner_radio_6ghz_ranges,
+			    u16 n_partner_6ghz_ranges)
+{
+	u16 i;
+	u16 n_local_this_6g_ranges = *n_this_6ghz_ranges;
+	struct ath12k_afc_freq_obj *merged_ranges;
+	const u16 n_merged_ranges = n_local_this_6g_ranges + n_partner_6ghz_ranges;
+
+	merged_ranges =
+	    krealloc(*this_radio_6ghz_ranges,
+		     n_merged_ranges * sizeof(struct ath12k_afc_freq_obj), GFP_ATOMIC);
+	if (!merged_ranges) {
+		ath12k_err(ab, "Failed to realloc merged_ranges\n");
+		return -ENOMEM;
+	}
+
+	if (merged_ranges[0].low_freq < partner_radio_6ghz_ranges[0].low_freq) {
+		memcpy(merged_ranges + n_local_this_6g_ranges,
+		       partner_radio_6ghz_ranges,
+		       n_partner_6ghz_ranges * sizeof(struct ath12k_afc_freq_obj));
+	} else {
+		memmove(merged_ranges + n_partner_6ghz_ranges, merged_ranges,
+			n_local_this_6g_ranges * sizeof(struct ath12k_afc_freq_obj));
+		memcpy(merged_ranges, partner_radio_6ghz_ranges,
+		       n_partner_6ghz_ranges * sizeof(struct ath12k_afc_freq_obj));
+	}
+
+	for (i = 0; i < n_merged_ranges; i++) {
+		ath12k_dbg(ab, ATH12K_DBG_AFC, "Combined AFC Range [%d] %d-%d\n",
+			   i, merged_ranges[i].low_freq, merged_ranges[i].high_freq);
+	}
+
+	*n_this_6ghz_ranges = n_merged_ranges;
+	*this_radio_6ghz_ranges = merged_ranges;
+
+	return 0;
+}
+
+/**
+ * ath12k_find_partner_ar_for_dual_6ghz_afc() - Check if any other radio
+ * in the same hardware has received the AFC payload.
+ * @ar: pointer to current ath12k
+ *
+ * Return: pointer to the partner radio if found, NULL otherwise
+ */
+static struct ath12k *
+ath12k_find_partner_ar_for_dual_6ghz_afc(struct ath12k *ar)
+{
+	struct ath12k *cur_ar = ar;
+	struct ath12k_hw *ah = ar->ah;
+	int i;
+
+	ar = ah->radio;
+	for (i = 0; i < ah->num_radio; i++, ar++) {
+		struct ath12k_base *ab;
+
+		if (ar == cur_ar)
+			continue;
+
+		if (!ar->supports_6ghz)
+			continue;
+
+		ab = ar->ab;
+		if (!ab->sp_rule || !ar->afc.is_6ghz_afc_power_event_received)
+			continue;
+
+		return ar;
+	}
+
+	return NULL;
+}
+
+/**
+ * ath12k_reg_intersect_with_partner_afc_info() - Intersect the current radio's
+ * AFC frequency ranges with the partner radio AFC frequency ranges.
+ * @ar: pointer to ath12k
+ * @intersected_afc_ranges: Input / Output pointer to the intersected AFC
+ * frequency ranges
+ * @num_intersected_ranges: Input / Output pointer to the number of AFC frequency
+ * ranges
+ *
+ * This API intersects the AFC frequency ranges with the partner radio
+ * AFC frequency ranges.
+ *
+ * Return: 0 on success, negative error code on failure
+ */
+static int
+ath12k_reg_intersect_with_partner_afc_info(struct ath12k *ar,
+					   struct ath12k_afc_freq_obj **intersected_afc_ranges,
+					   u16 *num_intersected_ranges)
+{
+	struct ath12k_base *ab = ar->ab;
+	struct ath12k_afc_freq_obj *partner_intersected_ranges;
+	u16 num_partner_intersected_ranges;
+	struct ath12k *partner_ar;
+	int num_partner_afc_rules;
+	int ret;
+
+	partner_ar = ath12k_find_partner_ar_for_dual_6ghz_afc(ar);
+	if (!partner_ar) {
+		ath12k_dbg(ab, ATH12K_DBG_AFC,
+			   "No Partner Radio found.\n");
+		return 0;
+	}
+
+	ath12k_dbg(ab, ATH12K_DBG_AFC,
+		   "Partner Radio has received AFC Resp\n");
+	spin_lock_bh(&partner_ar->data_lock);
+	num_partner_afc_rules = partner_ar->afc.afc_reg_info->num_freq_objs;
+	if (!num_partner_afc_rules) {
+		ath12k_dbg(ab, ATH12K_DBG_AFC,
+			   "No Partner AFC freq info\n");
+		spin_unlock_bh(&partner_ar->data_lock);
+		return 0;
+	}
+
+	partner_intersected_ranges =
+	    kzalloc(num_partner_afc_rules *
+		    sizeof(*partner_intersected_ranges), GFP_ATOMIC);
+
+	if (!partner_intersected_ranges) {
+		ath12k_err(ab, "Failed to alloc partner_intersected_ranges\n");
+		spin_unlock_bh(&partner_ar->data_lock);
+		return -ENOMEM;
+	}
+
+	ret = ath12k_reg_generate_valid_afc_ranges(partner_ar,
+						   partner_ar->afc.afc_reg_info,
+						   &partner_intersected_ranges,
+						   &num_partner_intersected_ranges);
+
+	if (ret) {
+		ath12k_warn(ab, "Failed to coalesce Partner AFC freq info\n");
+		spin_unlock_bh(&partner_ar->data_lock);
+		goto end;
+	}
+
+	spin_unlock_bh(&partner_ar->data_lock);
+
+	if (num_partner_intersected_ranges) {
+		if (!*num_intersected_ranges) {
+			*intersected_afc_ranges = partner_intersected_ranges;
+			*num_intersected_ranges =
+			    num_partner_intersected_ranges;
+
+			return 0;
+		}
+
+		ret = ath12k_concat_sorted_ranges(ab, intersected_afc_ranges,
+						  num_intersected_ranges,
+						  partner_intersected_ranges,
+						  num_partner_intersected_ranges);
+	}
+
+end:
+	kfree(partner_intersected_ranges);
+	return ret;
+}
+
 /**
  * ath12k_calculate_no_ir_ranges() - Calculate the No-IR ranges
  * @ar: pointer to ath12k
@@ -1897,6 +2072,14 @@ ath12k_reg_generate_afc_and_no_ir_ranges(struct ath12k *ar,
 		}
 	}
 
+	ret = ath12k_reg_intersect_with_partner_afc_info(ar, afc_ranges,
+							 num_afc_ranges);
+	if (ret) {
+		ath12k_warn(ab,
+			    "Failed to intersect with partner AFC info\n");
+		return ret;
+	}
+
 	if (!*num_afc_ranges) {
 		ath12k_warn(ab, "No AFC Freq Range Found\n");
 		/* Still update all freq range as NO_IR to clear prev payload */
@@ -2162,6 +2345,7 @@ ath12k_reg_build_new_regd_for_afc(struct ath12k *ar,
  * ath12k_reg_check_afc_payload_validity() - Check the validity of the AFC
  * payload.
  * @ar: pointer to ath12k
+ * @afc: pointer to the AFC payload information
  *
  * This API checks the validity of the AFC payload. This function also handles
  * Invalid / Empty payloads.
@@ -2172,11 +2356,12 @@ ath12k_reg_build_new_regd_for_afc(struct ath12k *ar,
  * Return: 0 on success, negative error code on failure
  */
 static int
-ath12_reg_check_afc_payload_validity(struct ath12k *ar)
+ath12_reg_check_afc_payload_validity(struct ath12k *ar,
+				     const struct ath12k_afc_info *afc)
 {
 	bool is_afc_chan_or_freq_obj_empty;
 	struct ath12k_base *ab = ar->ab;
-	const struct ath12k_afc_sp_reg_info *afc_reg_info = ar->afc.afc_reg_info;
+	const struct ath12k_afc_sp_reg_info *afc_reg_info = afc->afc_reg_info;
 
 	is_afc_chan_or_freq_obj_empty = (afc_reg_info->num_chan_objs == 0) &&
 					(afc_reg_info->num_freq_objs == 0);
@@ -2198,12 +2383,14 @@ ath12_reg_check_afc_payload_validity(struct ath12k *ar)
 		return 1;
 
 	/* Empty payload is a valid payload. So Mark Power event success */
+	ar->afc = *afc;
 	ar->afc.is_6ghz_afc_power_event_received = true;
 
 	return 1;
 }
 
-int ath12k_reg_process_afc_power_event(struct ath12k *ar)
+int ath12k_reg_process_afc_power_event(struct ath12k *ar,
+				       const struct ath12k_afc_info *afc)
 {
 	struct ieee80211_regdomain *new_regd = NULL;
 	struct ieee80211_regdomain *old_regd = NULL;
@@ -2215,21 +2402,29 @@ int ath12k_reg_process_afc_power_event(struct ath12k *ar)
 	u16 num_no_ir_ranges = 0;
 	int ret = 0, pdev_idx;
 
-	if (!ab->sp_rule || !ab->sp_rule->num_6ghz_sp_rule || !ar->afc.afc_reg_info) {
+	if (!ab->sp_rule || !ab->sp_rule->num_6ghz_sp_rule || !afc->afc_reg_info) {
 		ath12k_warn(ab, "SP rule not present or AFC payload is NULL\n");
 		return -EINVAL;
 	}
 
+	/* Make Sure that AFC power event from another radio is not proceessed
+	 * or cleared before we finish processing the current radio's AFC
+	 * power event.
+	 */
+	spin_lock_bh(&ah->afc_lock);
 	spin_lock_bh(&ar->data_lock);
 
-	if (ath12_reg_check_afc_payload_validity(ar)) {
+	if (ath12_reg_check_afc_payload_validity(ar, afc)) {
 		ret = -EINVAL;
 		goto end;
 	}
 
+	ath12k_free_afc_power_event_info(&ar->afc);
+	ath12k_dbg(ab, ATH12K_DBG_AFC, "New AFC info %pK\n", afc->afc_reg_info);
+	ar->afc = *afc;
 	ar->afc.is_6ghz_afc_power_event_received = true;
 
-	ret = ath12k_reg_generate_afc_and_no_ir_ranges(ar, ar->afc.afc_reg_info,
+	ret = ath12k_reg_generate_afc_and_no_ir_ranges(ar, afc->afc_reg_info,
 						       &afc_ranges,
 						       &num_afc_ranges,
 						       &no_ir_ranges,
@@ -2248,6 +2443,7 @@ int ath12k_reg_process_afc_power_event(struct ath12k *ar)
 	}
 
 	spin_unlock_bh(&ar->data_lock);
+	spin_unlock_bh(&ah->afc_lock);
 
 	spin_lock_bh(&ab->base_lock);
 	pdev_idx = ar->pdev_idx;
@@ -2263,6 +2459,7 @@ int ath12k_reg_process_afc_power_event(struct ath12k *ar)
 	return ret;
 end:
 	spin_unlock_bh(&ar->data_lock);
+	spin_unlock_bh(&ah->afc_lock);
 	kfree(afc_ranges);
 	kfree(no_ir_ranges);
 	return ret;
@@ -3283,22 +3480,21 @@ static int ath12k_reg_afc_start(struct ath12k_base *ab,
 
 void ath12k_free_afc_power_event_info(struct ath12k_afc_info *afc)
 {
-	struct ath12k *ar = container_of(afc, struct ath12k, afc);
 	struct ath12k_afc_sp_reg_info *afc_reg_info;
-	struct ath12k_afc_chan_obj *afc_chan_info;
-	struct ath12k_base *ab = ar->ab;
 	int num_chan_objs;
 	int i;
 
 	if (!afc->afc_reg_info)
 		return;
 
-	ath12k_dbg(ab, ATH12K_DBG_AFC, "Freeing afc info\n");
+	ath12k_dbg(NULL, ATH12K_DBG_AFC, "Freeing afc info %pK\n", afc->afc_reg_info);
 	afc_reg_info = afc->afc_reg_info;
 	num_chan_objs = afc_reg_info->num_chan_objs;
 	kfree(afc_reg_info->afc_freq_info);
 
 	for (i = 0; i < num_chan_objs; i++) {
+		struct ath12k_afc_chan_obj *afc_chan_info;
+
 		afc_chan_info = afc_reg_info->afc_chan_info + i;
 		kfree(afc_chan_info->chan_eirp_info);
 	}
@@ -3306,6 +3502,7 @@ void ath12k_free_afc_power_event_info(struct ath12k_afc_info *afc)
 	kfree(afc_reg_info->afc_chan_info);
 	kfree(afc_reg_info);
 	afc->afc_reg_info = NULL;
+	afc->is_6ghz_afc_power_event_received = false;
 }
 
 int ath12k_process_expiry_event(struct ath12k *ar)
@@ -3315,7 +3512,6 @@ int ath12k_process_expiry_event(struct ath12k *ar)
 
 	ath12k_dbg(ar->ab, ATH12K_DBG_AFC, "AFC expiry event subtype %d\n",
 		   afc->event_subtype);
-	afc->is_6ghz_afc_power_event_received = false;
 
 	switch (afc->event_subtype) {
 	case REG_AFC_EXPIRY_EVENT_START:
@@ -3329,7 +3525,9 @@ int ath12k_process_expiry_event(struct ath12k *ar)
 		break;
 	case REG_AFC_EXPIRY_EVENT_SWITCH_TO_LPI:
 		ret = ath12k_handle_invalid_afc_payload(ar);
+		spin_lock_bh(&ar->data_lock);
 		ath12k_free_afc_power_event_info(afc);
+		spin_unlock_bh(&ar->data_lock);
 		if (ret) {
 			ath12k_dbg(ar->ab, ATH12K_DBG_AFC, "Failed to process switch to LPI event\n");
 			return ret;
diff --git a/drivers/net/wireless/ath/ath12k/reg.h b/drivers/net/wireless/ath/ath12k/reg.h
index 562efd47a817..d9191824dfc4 100644
--- a/drivers/net/wireless/ath/ath12k/reg.h
+++ b/drivers/net/wireless/ath/ath12k/reg.h
@@ -573,7 +573,19 @@ int ath12k_regd_update(struct ath12k *ar, bool init);
 int ath12k_reg_update_chan_list(struct ath12k *ar, bool wait);
 int ath12k_reg_get_num_chans_in_band(struct ath12k *ar,
 				     struct ieee80211_supported_band *band);
-int ath12k_reg_process_afc_power_event(struct ath12k *ar);
+
+/**
+ * ath12k_reg_process_afc_power_event() - Process the AFC power event
+ * @ar: pointer to ath12k
+ * @afc: pointer to the AFC payload information
+ *
+ * This API processes the AFC power event and updates the regulatory domain
+ * accordingly.
+ *
+ * Return: 0 on success, negative error code on failure
+ */
+int ath12k_reg_process_afc_power_event(struct ath12k *ar,
+				       const struct ath12k_afc_info *afc);
 
 /**
  * ath12k_get_afc_req_info - Get AFC request information
diff --git a/drivers/net/wireless/ath/ath12k/vendor.c b/drivers/net/wireless/ath/ath12k/vendor.c
index 2197997a5379..dc868a7d2e18 100644
--- a/drivers/net/wireless/ath/ath12k/vendor.c
+++ b/drivers/net/wireless/ath/ath12k/vendor.c
@@ -769,7 +769,7 @@ ath12k_afc_power_event_update_or_get_len(struct ath12k *ar,
 	nla_nest_end_checked(vendor_event, nla_attr);
 
 	hw_idx = ath12k_get_radio_idx_by_freq(ar->ah->hw->wiphy,
-					      pwr_evt->afc_freq_info[0].low_freq + 10);
+					      ar->freq_range.start_freq);
 	if (vendor_event &&
 	    nla_put_u8(vendor_event,
 		       QCA_WLAN_VENDOR_ATTR_AFC_EVENT_HW_IDX, hw_idx)) {
@@ -982,7 +982,7 @@ ath12k_afc_expiry_event_update_or_get_len(struct ath12k *ar,
 	nla_nest_end_checked(vendor_event, nla_attr);
 
 	hw_idx = ath12k_get_radio_idx_by_freq(ar->ah->hw->wiphy,
-					      afc_req->freq_lst->range_objs[0].lowfreq + 10);
+					      ar->freq_range.start_freq);
 	if (vendor_event &&
 	    nla_put_u8(vendor_event,
 		       QCA_WLAN_VENDOR_ATTR_AFC_EVENT_HW_IDX, hw_idx)) {
@@ -1021,8 +1021,6 @@ int ath12k_send_afc_payload_reset(struct ath12k *ar)
 	int ret = -EINVAL;
 	int vendor_buffer_len, hw_index;
 	struct ath12k_base *ab = ar->ab;
-	struct ath12k_wmi_hal_reg_capabilities_ext_arg *reg_cap;
-	u16 freq_low;
 	struct ath12k_link_vif *tmp_arvif = NULL, *arvif;
 	struct wireless_dev *wdev;
 
@@ -1048,16 +1046,10 @@ int ath12k_send_afc_payload_reset(struct ath12k *ar)
 		goto out;
 	}
 
-	reg_cap = &ab->hal_reg_cap[ar->pdev_idx];
-	if (!reg_cap) {
-		ath12k_err(ab, "Unable to get reg cap for pdev %d\n", ar->pdev_idx);
-		goto out;
-	}
-
-	freq_low = reg_cap->low_5ghz_chan;
-	hw_index = ath12k_get_radio_idx_by_freq(ar->ah->hw->wiphy, freq_low);
+	hw_index = ath12k_get_radio_idx_by_freq(ar->ah->hw->wiphy, ar->freq_range.start_freq);
 	if (hw_index == -1) {
-		ath12k_err(ab, "Failed to get hw index for freq %d\n", freq_low);
+		ath12k_err(ab, "Failed to get hw index for freq %d\n", 
+			   ar->freq_range.start_freq);
 		goto out;
 	}
 
@@ -1321,9 +1313,11 @@ int ath12k_vendor_trigg_pri_link_migrate(struct wiphy *wiphy,
 	return ret;
 }
 
-int ath12k_vendor_send_power_update_complete(struct ath12k *ar)
+int
+ath12k_vendor_send_power_update_complete(struct ath12k *ar,
+					 struct ath12k_afc_info *afc)
 {
-	struct ath12k_afc_sp_reg_info *afc_reg_info = ar->afc.afc_reg_info;
+	struct ath12k_afc_sp_reg_info *afc_reg_info = afc->afc_reg_info;
 	struct ath12k_link_vif *tmp_arvif = NULL, *arvif;
 	struct sk_buff *vendor_event;
 	struct wireless_dev *wdev;
diff --git a/drivers/net/wireless/ath/ath12k/vendor.h b/drivers/net/wireless/ath/ath12k/vendor.h
index 4784d0750302..c9eb28b3c771 100644
--- a/drivers/net/wireless/ath/ath12k/vendor.h
+++ b/drivers/net/wireless/ath/ath12k/vendor.h
@@ -47,8 +47,13 @@ enum qca_nl80211_vendor_events {
  * vendor NL event to the application.
  *
  * @ar - Pointer to ar
+ * @afc - Pointer to AFC info
+ *
+ * Return: 0 on success, negative error code on failure
  */
-int ath12k_vendor_send_power_update_complete(struct ath12k *ar);
+int
+ath12k_vendor_send_power_update_complete(struct ath12k *ar,
+					 struct ath12k_afc_info *afc);
 
 /**
  * ath12k_send_afc_request - Send AFC request vendor NL event to the application
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index aa32de9fd9fd..65db565efc99 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -6911,8 +6911,9 @@ static int ath12k_wmi_afc_event_parser(struct ath12k_base *ab,
 	return ret;
 }
 
-static struct ath12k *ath12k_wmi_afc_process_fixed_param(struct ath12k_base *ab,
-							 void *ptr, size_t len)
+static struct ath12k *
+ath12k_wmi_afc_process_fixed_param(struct ath12k_base *ab,
+				   void *ptr, size_t len, struct ath12k_afc_info *afc)
 {
 	struct wmi_afc_event_fixed_param *fixed_param;
 	const struct wmi_tlv *tlv;
@@ -6947,9 +6948,7 @@ static struct ath12k *ath12k_wmi_afc_process_fixed_param(struct ath12k_base *ab,
 		return NULL;
 	}
 
-	ath12k_free_afc_power_event_info(&ar->afc);
-	memset(&ar->afc, 0, sizeof(ar->afc));
-	ar->afc.event_type = le32_to_cpu(fixed_param->event_type);
+	afc->event_type = le32_to_cpu(fixed_param->event_type);
 
 	return ar;
 }
@@ -6957,19 +6956,19 @@ static struct ath12k *ath12k_wmi_afc_process_fixed_param(struct ath12k_base *ab,
 static void ath12k_wmi_afc_event(struct ath12k_base *ab,
 				 struct sk_buff *skb)
 {
-	struct ath12k_afc_info *afc_info;
+	struct ath12k_afc_info afc = {};
+	struct ath12k_afc_info *afc_info = &afc;
 	struct ath12k *ar;
 	int ret;
 
-	ar = ath12k_wmi_afc_process_fixed_param(ab, skb->data, skb->len);
+	ar = ath12k_wmi_afc_process_fixed_param(ab, skb->data, skb->len, afc_info);
 	if (!ar) {
 		ath12k_warn(ab, "Failed to get ar for afc processing\n");
 		return;
 	}
 
-	afc_info = &ar->afc;
 	ath12k_dbg(ab, ATH12K_DBG_AFC, "Received AFC event of type %d for pdev: %d\n",
-		   ar->afc.event_type, ar->pdev->pdev_id);
+		   afc_info->event_type, ar->pdev->pdev_id);
 	ret = ath12k_wmi_tlv_iter(ab, skb->data, skb->len,
 				  ath12k_wmi_afc_event_parser, afc_info);
 	if (ret) {
@@ -6980,15 +6979,23 @@ static void ath12k_wmi_afc_event(struct ath12k_base *ab,
 	}
 	switch (afc_info->event_type) {
 	case ATH12K_AFC_EVENT_POWER_INFO:
-		ret = ath12k_reg_process_afc_power_event(ar);
+		ret = ath12k_reg_process_afc_power_event(ar, afc_info);
 		if (ret)
 			ath12k_warn(ab, "AFC reg rule update failed ret : %d\n",
 				    ret);
 
 		/* Update AFC application with power event update complete */
-		ath12k_vendor_send_power_update_complete(ar);
+		ath12k_vendor_send_power_update_complete(ar, afc_info);
+		if (afc_info->afc_reg_info->fw_status_code !=
+		    REG_FW_AFC_POWER_EVENT_SUCCESS) {
+			ath12k_free_afc_power_event_info(afc_info);
+		}
+
 		break;
 	case ATH12K_AFC_EVENT_TIMER_EXPIRY:
+		ar->afc.event_type = afc_info->event_type;
+		ar->afc.event_subtype = afc_info->event_subtype;
+		ar->afc.request_id = afc_info->request_id;
 		ret = ath12k_process_expiry_event(ar);
 		if (ret)
 			ath12k_warn(ab, "Failed to process expiry event\n");
-- 
2.34.1

